File: src/main/kotlin/codes/qdbp/serverplugin/listeners/PlayerDoorInteractListener.java
package codes.qdbp.serverplugin.listeners;

import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.BlockState;
import org.bukkit.block.data.Openable;
import org.bukkit.block.data.type.Door;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerInteractEvent;


public class PlayerDoorInteractListener implements Listener {

    @EventHandler
    public void onPlayerInteractWithDoor(PlayerInteractEvent event) {

        if (!(event.getAction().isRightClick())) return;
        if (event.getClickedBlock() == null) return;

        Block currentBlock = event.getClickedBlock();
        if (!(currentBlock.getType().toString().contains("DOOR"))) return;
        if (currentBlock.getType().toString().contains("IRON")) return;

        Door currentDoor = (Door) currentBlock.getBlockData();

        Block relativeBlock = currentBlock.getRelative(-1, 0, 0);
        Door relativeDoor;
        switch (currentDoor.getFacing()) {
            case NORTH -> {
                if (relativeBlock.getType().toString().matches(".*DOOR.*")) {
                } else if (currentBlock.getRelative(1, 0, 0).getType().toString().matches(".*DOOR.*")) {
                    relativeBlock = currentBlock.getRelative(1, 0, 0);
                } else {
                    return;
                }

                relativeDoor = (Door) relativeBlock.getBlockData();
                if (!(relativeDoor.getFacing().equals(BlockFace.NORTH))) return;
                if (relativeDoor.getHinge().equals(currentDoor.getHinge())) return;
                if (!(relativeDoor.getHalf().equals(currentDoor.getHalf()))) return;
            }
            case SOUTH -> {
                relativeBlock = currentBlock.getRelative(-1, 0, 0);

                if (relativeBlock.getType().toString().matches(".*DOOR.*")) {
                } else if (currentBlock.getRelative(1, 0, 0).getType().toString().matches(".*DOOR.*")) {
                    relativeBlock = currentBlock.getRelative(1, 0, 0);
                } else {
                    return;
                }


                relativeDoor = (Door) relativeBlock.getBlockData();
                if (!(relativeDoor.getFacing().equals(BlockFace.SOUTH))) return;
                if (relativeDoor.getHinge().equals(currentDoor.getHinge())) return;
                if (!(relativeDoor.getHalf().equals(currentDoor.getHalf()))) return;
            }
            case WEST -> {
                relativeBlock = currentBlock.getRelative(0, 0, -1);




                if (relativeBlock.getType().toString().matches(".*DOOR.*")) {
                } else if (currentBlock.getRelative(0, 0, 1).getType().toString().matches(".*DOOR.*")) {
                    relativeBlock = currentBlock.getRelative(0, 0, 1);
                } else {
                    return;
                }


                relativeDoor = (Door) relativeBlock.getBlockData();
                if (!(relativeDoor.getFacing().equals(BlockFace.WEST))) return;
                if (relativeDoor.getHinge().equals(currentDoor.getHinge())) return;
                if (!(relativeDoor.getHalf().equals(currentDoor.getHalf()))) return;

            }
            case EAST -> {
                relativeBlock = currentBlock.getRelative(0, 0, -1);


                if (relativeBlock.getType().toString().matches(".*DOOR.*")) {
                } else if (currentBlock.getRelative(0, 0, 1).getType().toString().matches(".*DOOR.*")) {
                    relativeBlock = currentBlock.getRelative(0, 0, 1);
                } else {
                    return;
                }

                relativeDoor = (Door) relativeBlock.getBlockData();
                if (!(relativeDoor.getFacing().equals(BlockFace.EAST))) return;
                if (relativeDoor.getHinge().equals(currentDoor.getHinge())) return;
                if (!(relativeDoor.getHalf().equals(currentDoor.getHalf()))) return;
            }
        }

        BlockState relativeBlockState = relativeBlock.getState();
        Openable relativeDoorOpenable = (Openable) relativeBlockState.getBlockData();


        Openable currentDoorOpenable = (Openable) currentBlock.getBlockData();
        if (currentDoorOpenable.isOpen()) {
            relativeDoorOpenable.setOpen(false);
            relativeBlockState.setBlockData(relativeDoorOpenable);
            relativeBlockState.update();
        } else if (!currentDoorOpenable.isOpen()) {
            relativeDoorOpenable.setOpen(true);
            relativeBlockState.setBlockData(relativeDoorOpenable);
            relativeBlockState.update();
        }
    }
}

File: src/main/kotlin/codes/qdbp/serverplugin/listeners/PlayerDoorInteractListener.kt
package codes.qdbp.serverplugin.listeners

import net.kyori.adventure.text.Component
import org.bukkit.Bukkit
import org.bukkit.block.Block
import org.bukkit.block.data.type.Door
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.block.BlockFace.*
import org.bukkit.block.ShulkerBox
import org.bukkit.block.data.Openable
import org.bukkit.event.inventory.InventoryType
import org.bukkit.inventory.meta.BlockStateMeta

class PlayerDoorInteractListener : Listener {

    @EventHandler
    fun onPlayerDoorInteract(event: PlayerInteractEvent) {


        if (event.player.inventory.itemInMainHand.type.toString().contains("SHULKER_BOX")) {

            if (event.action.isRightClick && event.clickedBlock?.type == null) {
                val shulkerMeta = event.player.inventory.itemInMainHand.itemMeta as BlockStateMeta
                val shulker = shulkerMeta.blockState as ShulkerBox
                val inventory = Bukkit.createInventory(event.player, InventoryType.SHULKER_BOX, Component.text("Shulker Manual"))
                inventory.contents = shulker.inventory.contents
                event.player.openInventory(inventory)
            }
        }

        if (!(event.action.isRightClick)) return
        val clickedBlock: Block = event.clickedBlock ?: return
        if (!(clickedBlock.type.toString().contains("DOOR"))) return
        if (clickedBlock.type.toString().contains("TRAP")) return
        if (clickedBlock.type.toString().contains("IRON")) return

        val clickedDoor = clickedBlock.blockData as Door
        var relativeBlock: Block
        val relativeDoor: Door

        when (clickedDoor.facing) {
            NORTH, SOUTH -> {
                relativeBlock = clickedBlock.getRelative(-1, 0, 0)
                when {
                    relativeBlock.type.toString().matches(Regex(".*DOOR*.")) -> {}
                    clickedBlock.getRelative(1, 0, 0).type.toString().matches(Regex(".*DOOR*.")) -> {
                        relativeBlock = clickedBlock.getRelative(1, 0, 0)
                    }
                    else -> return
                }
                relativeDoor = relativeBlock.blockData as Door
                if (relativeDoor.facing == WEST || relativeDoor.facing == EAST) return
                if (relativeDoor.hinge == clickedDoor.hinge) return
                if (relativeDoor.half != clickedDoor.half) return
            }

            WEST, EAST -> {
                relativeBlock = clickedBlock.getRelative(0, 0, -1)
                when {
                    relativeBlock.type.toString().matches(Regex(".*DOOR*.")) -> {}
                    clickedBlock.getRelative(0, 0, 1).type.toString().matches(Regex(".*DOOR*.")) -> {
                        relativeBlock = clickedBlock.getRelative(0, 0, 1)
                    }
                    else -> return
                }

                relativeDoor = relativeBlock.blockData as Door
                if (relativeDoor.facing == NORTH || relativeDoor.facing == SOUTH) return
                if (relativeDoor.hinge == clickedDoor.hinge) return
                if (relativeDoor.half != clickedDoor.half) return
            }
            else -> return
        }

        val relativeBlockState = relativeBlock.state
        val relativeDoorOpenable = relativeBlockState.blockData as Openable
        val clickedDoorOpenable = clickedBlock.blockData as Openable

        if (clickedDoorOpenable.isOpen) {
            relativeDoorOpenable.isOpen = false
            relativeBlockState.blockData = relativeDoorOpenable
            relativeBlockState.update()
        } else {
            relativeDoorOpenable.isOpen = true
            relativeBlockState.blockData = relativeDoorOpenable
            relativeBlockState.update()
        }

    }
}